### ISSUE TITLE
[BUG] After deleting the only VM of LB, LB controller reports following error

### ISSUE THREAD

**Issue by w13915984028** [harvester/harvester#5137]
**Describe the bug**

**To Reproduce**
1. Create a IPPool
2. Create one VM
3. Create one LB with workload type VM and selecting the VM in step 2
4. After a few minutes, delete the VM
5. In LB POD, such errors are continuously reported

> time="2024-02-08T16:54:39Z" level=error msg="error syncing 'default/vm1': handler harvester-lb-vmi-controller: admission webhook \"virtualmachineinstances-update-validator.kubevirt.io\" denied the request: update of VMI object is restricted, requeuing"

**Expected behavior**

**Support bundle**

```
LB with selected VM

# kubectl get loadbalancer -A
NAMESPACE   NAME   DESCRIPTION   WORKLOADTYPE   IPAM   ADDRESS       AGE
default     lb1                  vm             pool   192.168.5.5   45s

# kubectl get loadbalancer -A -oyaml
apiVersion: v1
items:
- apiVersion: loadbalancer.harvesterhci.io/v1beta1
  kind: LoadBalancer
  metadata:
    annotations:
      loadbalancer.harvesterhci.io/namespace: default
      loadbalancer.harvesterhci.io/network: ""
      loadbalancer.harvesterhci.io/project: ""
    creationTimestamp: "2024-02-08T16:56:48Z"
    finalizers:
    - wrangler.cattle.io/harvester-lb-controller
    generation: 2
    name: lb1
    namespace: default
    resourceVersion: "80926"
    uid: 9ea1730d-e933-49c2-9e02-8f855ca8f206
  spec:
    backendServerSelector:
      harvesterhci.io/ippool:
      - test
    healthCheck: {}
    ipPool: test-pool
    ipam: pool
    listeners:
    - backendPort: 80
      name: http
      port: 8080
      protocol: TCP
    workloadType: vm
  status:
    address: 192.168.5.5
    allocatedAddress:
      gateway: 192.168.5.1
      ip: 192.168.5.5
      ipPool: test-pool
      mask: 255.255.255.0
    backendServers:
    - 10.52.0.135
    conditions:
    - lastUpdateTime: "2024-02-08T16:56:49Z"
      status: "True"
      type: Ready
kind: List
metadata:
  resourceVersion: ""

 # kubectl get ippool.loadbalancer.harvesterhci.io -A -oyaml
apiVersion: v1
items:
- apiVersion: loadbalancer.harvesterhci.io/v1beta1
  kind: IPPool
  metadata:
    annotations:
      field.cattle.io/description: pool for test
    creationTimestamp: "2024-02-08T16:46:10Z"
    finalizers:
    - wrangler.cattle.io/harvester-ipam-controller
    generation: 3
    labels:
      loadbalancer.harvesterhci.io/global-ip-pool: "false"
      loadbalancer.harvesterhci.io/vid: "0"
    name: test-pool
    resourceVersion: "80898"
    uid: 03dac22e-7543-4f44-a5d0-1ace2936b596
  spec:
    ranges:
    - gateway: 192.168.5.1
      rangeEnd: 192.168.5.6
      rangeStart: 192.168.5.5
      subnet: 192.168.5.0/24
    selector: {}
  status:
    allocated:
      192.168.5.5: default/lb1
    available: 1
    conditions:
    - lastUpdateTime: "2024-02-08T16:46:10Z"
      status: "True"
      type: Ready
    lastAllocated: 192.168.5.5
    total: 2
kind: List
metadata:
  resourceVersion: ""

// when the vm is deleted, LB & IPPool have no change

```

**Note, to verify this bug, please focus on this phenomenon.**

With a further test, when creating a VM and there is no ippool/LB, the LB controller still has such message:
```
time="2024-02-12T15:06:48Z" level=error msg="error syncing 'default/vm1': handler harvester-lb-vmi-controller: admission webhook \"virtualmachineinstances-update-validator.kubevirt.io\" denied the request: update of VMI object is restricted, requeuing"
time="2024-02-12T15:06:49Z" level=error msg="error syncing 'default/vm1': handler harvester-lb-vmi-controller: admission webhook \"virtualmachineinstances-update-validator.kubevirt.io\" denied the request: update of VMI object is restricted, requeuing"
```

**Environment**
 - Harvester ISO version: V1.3.0-rc2
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): KVM VM based
 
 Observed when debugging #5033 

**Additional context**

**Comment by w13915984028** [comment]
With a further test, when creating a VM and there is no ippool/LB, the LB controller still has such message:
```
time="2024-02-12T15:06:48Z" level=error msg="error syncing 'default/vm1': handler harvester-lb-vmi-controller: admission webhook \"virtualmachineinstances-update-validator.kubevirt.io\" denied the request: update of VMI object is restricted, requeuing"
time="2024-02-12T15:06:49Z" level=error msg="error syncing 'default/vm1': handler harvester-lb-vmi-controller: admission webhook \"virtualmachineinstances-update-validator.kubevirt.io\" denied the request: update of VMI object is restricted, requeuing"
```

The direct cause is:
LB has a controller to watch VM OnRemove:
```
func (h *Handler) OnChange(_ string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
```

The wrangler framework will try to add a finalizer `wrangler.cattle.io/harvester-lb-vmi-controller` to VMI object, but this is denied.

The root cause is TBD.

After adding the finalizer manually, the message is gone:
```
- apiVersion: kubevirt.io/v1
  kind: VirtualMachineInstance
  metadata:
    annotations:
      harvesterhci.io/sshNames: '[]'
      kubevirt.io/latest-observed-api-version: v1
      kubevirt.io/storage-observed-api-version: v1
      kubevirt.io/vm-generation: "2"
    creationTimestamp: "2024-02-12T15:06:44Z"
    finalizers:
    - kubevirt.io/virtualMachineControllerFinalize
    - foregroundDeleteVirtualMachine
    - wrangler.cattle.io/VMIController.UnsetOwnerOfPVCs
    - wrangler.cattle.io/harvester-lb-vmi-controller
```

**Comment by w13915984028** [comment]
In Harvester v1.3.0, the kubevirt is v1.1.0, but the LB still uses kubevirt v0.5.4 as the client, which caused the above error; after go mod bump, the error message is gone. wrangler can add new items to finalizers correctly. 

```
- apiVersion: kubevirt.io/v1
  kind: VirtualMachineInstance
  metadata:
    annotations:
      harvesterhci.io/sshNames: '[]'
      kubevirt.io/latest-observed-api-version: v1
      kubevirt.io/storage-observed-api-version: v1
      kubevirt.io/vm-generation: "2"
    creationTimestamp: "2024-03-08T17:38:55Z"
    finalizers:
    - kubevirt.io/virtualMachineControllerFinalize
    - foregroundDeleteVirtualMachine
    - wrangler.cattle.io/VMIController.UnsetOwnerOfPVCs
    - wrangler.cattle.io/harvester-lb-vmi-controller
    generation: 9
    labels:
      harvesterhci.io/vmName: vm2
      kubevirt.io/nodeName: harv41

$ kubect logs -n harvester-system harvester-load-balancer-6754b798bb-hjptw
I0308 17:37:16.890940       1 leaderelection.go:248] attempting to acquire leader lease kube-system/harvester-load-balancer...
I0308 17:37:16.905632       1 leaderelection.go:258] successfully acquired lease kube-system/harvester-load-balancer
time="2024-03-08T17:37:16Z" level=info msg="Initialize IP pool from kube-vip configmap"
time="2024-03-08T17:37:17Z" level=info msg="Starting /v1, Kind=Service controller"
time="2024-03-08T17:37:17Z" level=info msg="Starting discovery.k8s.io/v1, Kind=EndpointSlice controller"
time="2024-03-08T17:37:17Z" level=info msg="Starting k8s.cni.cncf.io/v1, Kind=NetworkAttachmentDefinition controller"
time="2024-03-08T17:37:17Z" level=info msg="Starting loadbalancer.harvesterhci.io/v1beta1, Kind=IPPool controller"
time="2024-03-08T17:37:17Z" level=info msg="Starting loadbalancer.harvesterhci.io/v1beta1, Kind=LoadBalancer controller"
time="2024-03-08T17:37:17Z" level=info msg="Starting kubevirt.io/v1, Kind=VirtualMachineInstance controller"
```

**Comment by bk201** [comment]
@w13915984028 Is this a blocker to v1.3.0?

**Comment by w13915984028** [comment]
After PR https://github.com/harvester/load-balancer-harvester/pull/26 is merged, a new `master-head`  image of `harvester-load-balancer` is built.

The harvester/harvester-installer will always use the `master-head` image of `harvester-load-balancer` to build-iso. With a latest local build ISO, the issue is verifyed to be solved.

https://github.com/harvester/harvester/blob/a2c98e9606eec79c839c5bed095cc32fc1616b3c/deploy/charts/harvester/values.yaml#L453
```
harvester-load-balancer:
  enabled: false
  image:
    imagePullPolicy: Always
    repository: rancher/harvester-load-balancer
    tag: master-head
  webhook:
    image:
      imagePullPolicy: IfNotPresent
      repository: rancher/harvester-load-balancer-webhook
      tag: master-head
```

The `start/stop` vm will not lead to any error message in `harvester-load-balancer-*` pod.

```
$ kubectl logs -n harvester-system harvester-load-balancer-67f8b88d9c-5sm6c
I0312 21:53:30.753139       1 leaderelection.go:248] attempting to acquire leader lease kube-system/harvester-load-balancer...
I0312 21:53:30.774262       1 leaderelection.go:258] successfully acquired lease kube-system/harvester-load-balancer
time="2024-03-12T21:53:30Z" level=info msg="Initialize IP pool from kube-vip configmap"
time="2024-03-12T21:53:30Z" level=info msg="Starting /v1, Kind=Service controller"
time="2024-03-12T21:53:31Z" level=info msg="Starting discovery.k8s.io/v1, Kind=EndpointSlice controller"
time="2024-03-12T21:53:31Z" level=info msg="Starting k8s.cni.cncf.io/v1, Kind=NetworkAttachmentDefinition controller"
time="2024-03-12T21:53:31Z" level=info msg="Starting loadbalancer.harvesterhci.io/v1beta1, Kind=IPPool controller"
time="2024-03-12T21:53:31Z" level=info msg="Starting loadbalancer.harvesterhci.io/v1beta1, Kind=LoadBalancer controller"
time="2024-03-12T21:53:31Z" level=info msg="Starting kubevirt.io/v1, Kind=VirtualMachineInstance controller"
```

**Comment by w13915984028** [comment]
In Harvester v130 master-head, this issue is ready to be validated.
In Harvester v130-rc*, it needs to wait unti the PR https://github.com/harvester/harvester/pull/5354 is merged and a new RC is released.

**Comment by irishgordo** [comment]
@w13915984028 

On Harvester single-node Version: v1.3-7cb8315c-head 
I edited both deployments of:
- load-balancer
- load-balancer-webhook

And set them both to be at `master-head` tag of their respective images.
Watched it roll the replica sets.
And the new LB content is there.

I validated that given a file-server VM built on the 10.20.X.Y Masquerade net, with a label that is leveraged for a selector for a Load Balancer IP w/ IP Pool.
That when the VM is deleted, after running for some time, that there is no error messages present in the load balancer pod.

I'll go ahead and close this out as this looks good :smile: :+1: 

IP Pool based LB:


https://github.com/harvester/harvester/assets/5370752/1ee46e76-e2da-4fee-9cdd-ff728f2389f6



Also checked with DHCP:

https://github.com/harvester/harvester/assets/5370752/6704ad7a-4d78-49f2-b2e7-d3d44e3320e0
